### ISSUE TITLE
Figure out how to make Timoni's modules work with Flux

### ISSUE THREAD

**Issue by stefanprodan** [stefanprodan/timoni#10]
[Flux](https://fluxcd.io) source-controller is compatible with Timoni's modules published to container registries. The source-controller can pull the modules inside the cluster and verify their cosign signatures, what's missing is a dedicated controller that can managed Timoni's instances. One solution would to collaborate with @phoban01 and extend https://github.com/phoban01/cue-flux-controller with support for Timoni's instances, where the values would be supplied from Kubernetes Secrets and ConfigMaps. Another solution would be to create a timoni-controller from scratch.

**Comment by squaremo** [comment]
> One solution would be to [...] and extend https://github.com/phoban01/cue-flux-controller with support for Timoni's instances, 
> Another solution would be to create a timoni-controller from scratch.

The deciding factor might be this: are there any aspirations towards lifecycle management in Timoni? Things like Helm's pre-install and post-install hooks, I mean. If there are, this would end up being more like the Helm controller; but if not, I think it would be keeping in the spirit of the existing Cue controller to teach it about Timoni modules.

**Comment by stefanprodan** [comment]
> The deciding factor might be this: are there any aspirations towards lifecycle management in Timoni? Things like Helm's pre-install and post-install hooks

Yes that's a good point. Currently Timoni doesn't have the same hooks feature like Helm but it does have a feature to group Kubernetes manifests and it can perform a multi-step apply .e.g.

```
timoni: {

	// Pass Kubernetes resources outputted by the instance
	// to Timoni's multi-step apply.
	apply: prerequisites: [...]
	apply: frontend: [...]
	apply: backend: [...]
	apply: tests: [...]
}
```

Example here: https://github.com/stefanprodan/timoni/blob/main/examples/redis/timoni.cue#L36-L39

When multiple stages are specified within `timoni.apply`, Timoni applies them in order, and moves to the next stage only if the previous one becomes ready. A Flux controller for Timoni should have the same behaviour.

**Comment by onedr0p** [comment]
Maybe there could be a timoni-controller and a `TimoniRelease` CR that looks similar to the `HelmRelease` spec but built for specifically for Timoni?

phoban01/cue-flux-controller doesn't appear to be active (2 commits in the past year) so the project status is unclear and might cause problems on integrating timoni features?

**Comment by syncharny** [comment]
Is it only the phased apply that differentiates the flux pipeline from timoni apply? Could we have timoni build accept query paths so that we could store pre-deplomentsl specs in one folder, and post-deployment specs in another folder, then use flux's job pipeline?

https://fluxcd.io/flux/use-cases/running-jobs/

**Comment by emil-jacero** [comment]
@stefanprodan Is your approach to developing the runtime features and getting a feeling for the workflow before investing time into developing a flux controller or are there no plans for a controller at the moment?

**Comment by Nalum** [comment]
@emil-jacero there is an [RFC](https://github.com/stefanprodan/timoni/blob/spec-controller/docs%2Fcontroller%2Fspecs.md) in progress on what a controller might look like.

**Comment by artem-nefedov** [comment]
If Timoni is to replace Helm, I would at least expect the following (this goes for both Timoni itself and whatever controller will be managing it in cluster):

1. Ability to provide "values" based on schema defined in the module, similar to values with json schema in Helm chart, or variables in Terraform module.
2. Support for pre/post-action hooks. This part doesn't seem to be fully solvable by Timoni's application order, as I don't see how it can trigger pre-delete/post-delete Job for example.

**Comment by zs-dima** [comment]
@artem-nefedov I mentioned it in the other issue with TimoniRelease example:
https://github.com/stefanprodan/timoni/issues/296
However would be nice to see universal solution that will work with ArgoCD as well.

**Comment by salotz** [comment]
It looks like from the docs that timoni controller is planned, but just wanted to put in my use case.

I like the idea of being able to deploy the flux-aio into customer environments and have it track drift and update helm charts we deploy them (via the Timoni module). The same for Timoni modules would be great.

Because our installations are sort of hands off after we install them having drift detection (those pesky customers messing with things) and automatic updates straight from an OCI registry without customer intervention would probably handle a lot of support requests.

**Comment by salotz** [comment]
We've been using timoni with great success for most of this year. I'm just curious as to how much more dogfooding would be necessary to make the controller possible to focus on. In our experience there really hasn't been any issues with the interfaces at all. Only minor complaints is on some inconsistency in the CLI command hierarchy which I don't think effects the controller much.

**Comment by errordeveloper** [comment]
@salotz I think sponsoring the project might be an option to get this moving forward.

**Comment by salotz** [comment]
@errordeveloper that could be a possibility. However, it seemed there were other blockers I wanted to check in on.
